### PR TITLE
Update Architecture Diagram

### LINKED ISSUE #735: Update Architecture Diagram
The architecture diagrams need an update, post Trillian.

### PR DESCRIPTION
**Pull request by gdbelvin**
Fixes #735

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -2,78 +2,80 @@
...
+
+# Commitment Table
+The commitment table stores account values and the associated commitment key 
+nessesary to verify the commitment stored in the Trillian Map. 
+
+# Mutation Table
+When a user wishes to make a change to their account, they create a signed change
+request (also known as a mutation) and send it to a Key Transparency frontend.
+
+The frontend then saves the mutation in the mutation table, allowin the database
```
Comment (Liamsi) [inline docs/architecture.md]: typo: s/allowin/allowing

File: docs/architecture.md [hunk 2]
```diff
@@ -2,78 +2,80 @@
...
+# Commitment Table
+The commitment table stores account values and the associated commitment key 
+nessesary to verify the commitment stored in the Trillian Map. 
+
+# Mutation Table
+When a user wishes to make a change to their account, they create a signed change
+request (also known as a mutation) and send it to a Key Transparency frontend.
+
+The frontend then saves the mutation in the mutation table, allowin the database 
+to assign the mutation a monotinically increasing sequence number or timestamp,
```
Comment (Liamsi) [inline docs/architecture.md]: s/ monotinically/monotonically

File: docs/architecture.md [hunk 3]
```diff
@@ -2,78 +2,80 @@
...
+
+# Mutation Table
+When a user wishes to make a change to their account, they create a signed change
+request (also known as a mutation) and send it to a Key Transparency frontend.
+
+The frontend then saves the mutation in the mutation table, allowin the database 
+to assign the mutation a monotinically increasing sequence number or timestamp, 
+establishing an authoritative ordering for new mutations.
+
+This strict ordering requirement could be loostened in the future for
```
Comment (Liamsi) [inline docs/architecture.md]: loosened?

File: docs/architecture.md [hunk 4]
```diff
@@ -2,78 +2,80 @@
...
+the Trillian Log which must sequence the new SMR before the front ends will
+start using the new map revision. 
+
+After each new map revision, the sequencer will also send the new SignedMapRoot,
+SignedLogRoot, Mutations, and associated proofs to the front ends over a
+streaming gRPC channel. The frontends will then forward those same notification
+to active monitors over a streaming gRPC channel.
+
+# Mutation
+Mutations in Key Transparency are defined as a SignedKeyValue object.
```
Comment (Liamsi) [inline docs/architecture.md]: Not sure if the camel case isn't confusing here. The type is called `SignedKV`.

File: docs/architecture.md [hunk 5]
```diff
@@ -2,78 +2,80 @@
...
+streaming gRPC channel. The frontends will then forward those same notification
+to active monitors over a streaming gRPC channel.
+
+# Mutation
+Mutations in Key Transparency are defined as a SignedKeyValue object. 
+- The Key must the the valid index associated with the user's identifier. 
+- The Value is an object that contains 
+   - A cryptographic commitment to the user's data.
+   - The set of public keys that are allowed to update this account.
+- SignedKV also contains the hash of the previous SignedKeyValue. This helps
```
Comment (Liamsi) [inline docs/architecture.md]: Consistent naming: Use either SignedKV, or signed key-value, or SignedKeyValue for describing the thing.

File: docs/architecture.md [hunk 6]
```diff
@@ -2,78 +2,80 @@
...
+  break race conditions and it forms a hash chain in each account.
+
+# Monitors
+Monitors process and verify the mutations that make each new epoch.
+Monitors verify various policy properties of the SignedKeyValues in the
+Trillian Map.  In particular, monitors verify that 
+- Back pointers in each leaf do not skip over any values - an operation that
+  would be bandwidth intensive for mobile clients.  
+- SignedKeyValues are properly signed by public keys declared in the prior
+  epoch.
```
Comment (Liamsi) [inline docs/architecture.md]: Isn't it necessary that the SignedKV is also signed by the new keys (in the new epoch)?

File: docs/architecture.md [hunk 7]
```diff
@@ -2,78 +2,80 @@
...
+# Monitors
+Monitors process and verify the mutations that make each new epoch.
+Monitors verify various policy properties of the SignedKeyValues in the
+Trillian Map.  In particular, monitors verify that 
+- Back pointers in each leaf do not skip over any values - an operation that
+  would be bandwidth intensive for mobile clients.  
+- SignedKeyValues are properly signed by public keys declared in the prior
+  epoch. 
+
+Monitors also observe the Trillian Log to detect any log forks.
```
Comment (Liamsi) [inline docs/architecture.md]: This sounds like all monitors will have direct access to the log which isn't that case (at least for most monitors).

File: docs/architecture.md [hunk 8]
```diff
@@ -2,78 +2,80 @@
...
-Stores a signed list of mutations and epoch advancement markers that come out of 
-the queue.
-
-# Log of Signed Map Heads
-Stores a signed list of Signed Map Heads (SMHs), one for each epoch.
+#Clients
+Clients make requests to Key Transparency servers over HTTPS / JSON or gRPC.  
+
+# Key Transparency Server
+The front end servers reveal the mappings between user identifiers (eg email
```
Comment (Liamsi) [inline docs/architecture.md]: nit: e.g.
American style: , e.g.,
UK: without comma

Comment (gdbelvin) [inline docs/architecture.md]: thanks, done

File: docs/architecture.md (hunk 1 above)
Comment (gdbelvin) [inline docs/architecture.md]: done

File: docs/architecture.md (hunk 2 above)
Comment (gdbelvin) [inline docs/architecture.md]: done

File: docs/architecture.md (hunk 3 above)
Comment (gdbelvin) [inline docs/architecture.md]: relaxed

File: docs/architecture.md (hunk 5 above)
Comment (gdbelvin) [inline docs/architecture.md]: done

Comment (gdbelvin) [inline docs/architecture.md]: We should also probably rename this struct to properly be "mutation" in our protos.

File: docs/architecture.md (hunk 7 above)
Comment (gdbelvin) [inline docs/architecture.md]: done

File: docs/architecture.md (hunk 5 above)
Comment (Liamsi) [inline docs/architecture.md]: I like this idea. We would need to find another name for [`Mutation`](https://github.com/google/keytransparency/blob/8eb3a9af4a8f06bfd4c316a2ed1b8c77a73cb336/core/proto/keytransparency_v1_types/keytransparency_v1_types.proto#L101) which contains a `SignedKV`.

File: docs/architecture.md [hunk 9]
```diff
@@ -2,78 +2,81 @@
...
+- The mutation also contains the hash of the previous mutation. This helps
+  break race conditions and it forms a hash chain in each account.
+
+# Monitors
+Monitors process and verify the mutations that make each new epoch.
+Monitors verify various policy properties of the signed key-values in the
+Trillian Map.  In particular, monitors verify that 
+- Back pointers in each leaf do not skip over any values - an operation that
+  would be bandwidth intensive for mobile clients.  
+- signed key-values are properly signed by public keys declared in the prior
```
Comment (Liamsi) [inline docs/architecture.md]: nit: they are actually signed by the private keys and verified using the public keys that they were signed.

**Pull request merged**